### ISSUE TITLE
[BUG] Don't auto prune *.tar.gz files and don't remove the compatibility tools from the UI

### ISSUE THREAD

**Issue by wehagy** [Vysp3r/ProtonPlus#43]
**Describe the bug**
If I click to download a compatibility tool the ProtonPlus download and extract, but don't auto prune *.tar.gz file and I can't remove the tool using the UI.

**To Reproduce**
1. Launch ProtonPlus from terminal.
```
~/.var/app/com.valvesoftware.Steam/data/Steam/compatibilitytools.d$ flatpak run com.vysp3r.ProtonPlus
Trash portal failed on /home/wgimenes/.var/app/com.vysp3r.ProtonPlus/config/preferences.json
Error opening file “/home/wgimenes/.var/app/com.vysp3r.ProtonPlus/config/preferences.json”: File exists
```
2. Click to install new compatibility tool on flatpak Steam and finish installation.
```
Trash portal failed on /home/wgimenes/.var/app/com.valvesoftware.Steam/data/Steam/compatibilitytools.d/GE-Proton7-42.tar.gz
```
3. Click to remove the compatibility tool and click Yes.
```
Trash portal failed on /home/wgimenes/.var/app/com.valvesoftware.Steam/data/Steam/compatibilitytools.d/GE-Proton7-42
```
4. See contents of the Steam compatibilitytools.d folder.
```
~/.var/app/com.valvesoftware.Steam/data/Steam/compatibilitytools.d$ ls -l
total 405004
drwxrwxr-x 1 wgimenes wgimenes       306 dez  2 20:22 GE-Proton7-42
-rw-r--r-- 1 wgimenes wgimenes 414720148 dez  2 21:14 GE-Proton7-42.tar.gz
```

**Expected behavior**
ProtonPlus auto prune *.tar.gz files and delete the compatibility tools from the UI.

**Desktop (please complete the following information):**
 - OS: Fedora 37, Manjaro, Debian bookworm/testing
 - DE: GNOME 43
 - Type: Flatpak

**Additional context**
Builded using flatpak builder based on branch [https://github.com/Vysp3r/ProtonPlus/tree/37-bugcritical-error-the-application-crashes-during-startup](url)

**Comment by Vysp3r** [comment]
I should be able to fix that this weekend.

**Comment by Vysp3r** [comment]
I'll probably do it later today, but can't confirm. Anyway even if I do it today, the next update will not be today.

**Comment by Vysp3r** [comment]
Would you be able to test from the source from the branch associated with this issue and tell me if it's working for you?

**Comment by wehagy** [comment]
Tested but I need to patch with branch [37-bugcritical-error-the-application-crashes-during-startup](https://github.com/Vysp3r/ProtonPlus/tree/37-bugcritical-error-the-application-crashes-during-startup).

The *.tar.gz file is removed fine, but now takes so much time to remove some compatibility tools, like 30s-50s and i see the message "com.vysper.ProtonPlus is not responding", the file is removed anyway, just takes time to finish.


![Screenshot from 2022-12-18 22-26-30](https://user-images.githubusercontent.com/95216845/208331263-594008b6-64e7-4716-8d7b-c4d7de271dae.png)


I downloaded several tools and try the option "Clean launcher", the same happens, all files is removed just takes time to finish.


![Screenshot from 2022-12-18 22-40-48](https://user-images.githubusercontent.com/95216845/208331944-90892b84-c910-4690-bd89-0bfda5db1b46.png)

**Comment by Vysp3r** [comment]
Oh I forgot to put that in another thread, my bad. For the perfomance problem, that's to be expected since I needed to write a recursive function to delete everything which may not be the most optimized way, but I can't do much more. Personally it took no time to delete them tho. What's your system specs if I may?

**Comment by wehagy** [comment]
> I had to create that recursive function since you did not support the trash function for I don't know what reasons.

And my 3 systems have the same problem, just don't know why too.

**Comment by Vysp3r** [comment]
The problem is the only way left for me to do something to delete file in a fast way is to run a command which is a bit meh in my book. If I remember correctly some people had problem with that too. So we are a bit in an impasse.

**Comment by Vysp3r** [comment]
I've looked an I can't find any mention of that causing a problem, but I'm sure I read that somewhere.

**Comment by Vysp3r** [comment]
Can you tell me if it works for you now? I've gone the command way instead of my custom delete function.

**Comment by wehagy** [comment]
---


### Download and delete Boxtron/Roberta OK


---


### Download and delete Proton-GE/NorthststarProton NOK.

\
ProtonPlus is not responding, need click to force close.

Spam forever until I force to close.


```
(protonplus:2): Gtk-CRITICAL **: 01:17:11.420: gtk_widget_get_visible: assertion 'GTK_IS_WIDGET (widget)' failed

(protonplus:2): Gtk-CRITICAL **: 01:17:11.420: gtk_widget_get_visible: assertion 'GTK_IS_WIDGET (widget)' failed

(protonplus:2): Gtk-CRITICAL **: 01:17:11.420: gtk_widget_get_visible: assertion 'GTK_IS_WIDGET (widget)' failed

(protonplus:2): Gtk-CRITICAL **: 01:17:11.420: gtk_widget_get_visible: assertion 'GTK_IS_WIDGET (widget)' failed

(protonplus:2): Gtk-CRITICAL **: 01:17:11.420: gtk_widget_get_visible: assertion 'GTK_IS_WIDGET (widget)' failed

(protonplus:2): Gtk-CRITICAL **: 01:17:11.420: gtk_widget_get_visible: assertion 'GTK_IS_WIDGET (widget)' failed

(protonplus:2): Gtk-CRITICAL **: 01:17:11.420: gtk_widget_get_visible: assertion 'GTK_IS_WIDGET (widget)' failedApplication exited
```

\
Sometimes I see this on output and ProtonPlus freeze until I close.


```
(protonplus:2): Gtk-WARNING **: 01:32:03.308: Trying to snapshot ProtonPlusWindowsHome 0x55a13cf442d0 without a current allocation
Application exited
```


---


### Download and delete Luxorpeda NOK

\
ProtonPlus is not responding, need click to force close.

```
Aplication exited
```


---


In all situations the compatibility tool and *.tar.gz files is removed and bulk remove is working very fine.

**Comment by Vysp3r** [comment]
I guess I'll have to do it with my recursive function then. I'll try to improve it.

**Comment by wehagy** [comment]
The recursive function is ok, just a little slow for some reason, it's annoying that Vala doesn't have a "delete folder anyway" function.

**Comment by wehagy** [comment]
No rush, and now I know why trash function not work (maybe), compiled the source code and run (out of curiosity I compiled it in fedora 37 and ran it in debian bookworm/testing), and now I see the error.


```
Trashing on system internal mounts is not supported
```

\
Found it's because of "bind mounts", the relevant line in my fstab.


```
LABEL=RAID1-HD /home/wgimenes/.var btrfs nosuid,nodev,compress-force=zstd,user_subvol_rm_allowed,autodefrag,x-systemd.device-timeout=0,subvolid=265,subvol=/USERDATA/wgimenes/.var 0 0
```

\
Basically I mount flatpak app settings from a different storage than my root filesystem, now that I realized that Nautilus/Files doesn't have the option to send it to the trash (I just rm -rf or shift+del all the time, I have snapshots, not reason to use trash).


![Screenshot from 2022-12-19 23-27-24](https://user-images.githubusercontent.com/95216845/208567706-c1387d6f-810c-42e6-b38c-d6640c0eccf6.png)

\
The performance hit is because that folder is on my HDD's, I didn't mention it but all my files are also encrypted, and sorry I know my english is not good but i'm trying my best :cry:.

**Comment by Vysp3r** [comment]
Dont worry, your english is better then some people I know ahah

**Comment by Vysp3r** [comment]
In that case I'm thinking I can use the trash function and if it doesnt work I use my recursive function as a fallback method. Of course I will improve it so its not that slow.

**Comment by wehagy** [comment]
It's a fair solution, maybe most people don't even face this "bug" in the first place.

**Comment by Vysp3r** [comment]
Yeah I think most people have access to the trash ahah

**Comment by Vysp3r** [comment]
Can you test the branch from this issue please :)
I decided not to make my function faster since it's a lot of multi threading stuff and my I'm not the best with that. I'll do it for next update, but for now I think pushing an update that fixes this issue and some others is more important. Don't worry tho, I will not forget you tho. For 0.3.3 it will be better ;)

**Comment by Vysp3r** [comment]
Oh damn, yeah I should do that. Thanks 😊

**Comment by Vysp3r** [comment]
Okay so now I did what you said. Tell me if it's any better.

**Comment by Vysp3r** [comment]
Oh wait, I need to fix something else. Sorry ahah

**Comment by Vysp3r** [comment]
Okay, so basically I did like you told me and I made so that the cleanup utility creates a thread for each folder to delete instead of a thread for all of them so it should be way faster to delete using that since they will all be deleted at the same time.

**Comment by Vysp3r** [comment]
For people that can use the trash() function it will change nothing, but for you it should be a big plus.

**Comment by Vysp3r** [comment]
Could you test and tell me if it's working for you :)

**Comment by Vysp3r** [comment]
IT WORKS!

**Comment by Vysp3r** [comment]
Tell me if that works for you and if so I'm gonna publish an update that is really due ahah

**Comment by wehagy** [comment]
Oh god, so many weird things happening right now, the try "trash" and after try "delete" is vanished, good work mate :+1:
The *.tar.gz files as always works, and most of the time compatibility tools are removed (single or bulk) without any problem, but it still persists in taking time to remove the compatibility tools, sometimes the app crashes if I try to remove something, and I see that on the output.

\
Spamming until I force close the app.

```
...

(protonplus:2): Gtk-CRITICAL **: 21:50:12.979: gtk_widget_get_visible: assertion 'GTK_IS_WIDGET (widget)' failed

(protonplus:2): Gtk-CRITICAL **: 21:50:12.979: gtk_widget_get_visible: assertion 'GTK_IS_WIDGET (widget)' failed

(protonplus:2): Gtk-CRITICAL **: 21:50:12.979: gtk_widget_get_visible: assertion 'GTK_IS_WIDGET (widget)' failed

(protonplus:2): Gtk-CRITICAL **: 21:50:12.979: gtk_widget_get_visible: assertion 'GTK_IS_WIDGET (widget)' failed

(protonplus:2): Gtk-CRITICAL **: 21:50:12.979: gtk_widget_get_visible: assertion 'GTK_IS_WIDGET (widget)' failed

(protonplus:2): Gtk-CRITICAL **: 21:50:12.979: gtk_widget_get_visible: assertion 'GTK_IS_WIDGET (widget)' failed
Application exited
```

\
The same as above but not spam and the app closes without any interaction.

```
(protonplus:2): Gtk-CRITICAL **: 21:52:44.691: gtk_widget_get_visible: assertion 'GTK_IS_WIDGET (widget)' failed

(protonplus:2): Gtk-CRITICAL **: 21:52:44.691: gtk_widget_get_visible: assertion 'GTK_IS_WIDGET (widget)' failed
Application exited
```

\
App freezes and I need to force close, and partially removed the compatibility tool, rarely occurs.

```
Error removing file /home/wgimenes/.var/app/com.valvesoftware.Steam/data/Steam/compatibilitytools.d/GE-Proton7-43/files: Directory not empty
Error removing file /home/wgimenes/.var/app/com.valvesoftware.Steam/data/Steam/compatibilitytools.d/GE-Proton7-43: Directory not empty

(protonplus:2): GLib-CRITICAL **: 23:05:46.602: g_sequence_get: assertion '!is_end (iter)' failed
Application exited
```

\
I thought if I moved the steam compatibility tool folder to an SSD if it would improve performance, I made a btrfs subvolume on the SSD and mounted the subvolume in the Steam compatibility tool folder, checked with Nautilus/Files that the recycle bin does not exist, tested again, unfortunately got the same results :cry:

**Comment by Vysp3r** [comment]
Ffs, I think I will break my keyboard ahah

**Comment by Vysp3r** [comment]
Updated the branch from this issue. I made it cleaner and a bit simpler. Tell me if that works any better for you.

**Comment by wehagy** [comment]
MY GOD, GOOD NEWS, the long time to remove is here unfortunately, but have been improved a lot, now it takes to remove one compatibility tool 18s-22s, reduced by almost half the time, I am very happy with the results, the bulk deletion still takes a long time, but I ended up not timing it before, so I can't confirm, but it also seems to have improved.

The bad news is the app crashes sometimes, and if crashes takes the same time to remove the compatibility tools like before, this is rare to happen.

```
(protonplus:2): Gtk-CRITICAL **: 03:26:37.758: gtk_widget_get_visible: assertion 'GTK_IS_WIDGET (widget)' failed
Application exited
```

Great job mate you deserve it :clap:

**Comment by Vysp3r** [comment]
I will kill that widget once I find him ahah.

**Comment by Vysp3r** [comment]
I'll continue working on a fix for that tomorrow when I wake up.

**Comment by Vysp3r** [comment]
I tried to optimize the process a bit, can you tell me if now that works any better please :)

**Comment by Vysp3r** [comment]
I just closed another issue where GLib was the issue. I think I will try to use Posix to delete stuff to for people in your position.

**Comment by Vysp3r** [comment]
Since this issue was fixed in #81, I will now close it.
